hkmc2/shared/src/main/scala/hkmc2/bbml/bbML.scala

```diff
@@ -208,12 +210,18 @@ class BBTyper(using elState: Elaborator.State, tl: TL):
   private def typeType(ty: Term)(using ctx: Ctx): GeneralType =
     typeAndSubstType(ty, pol = true)(using Map.empty)
   
-  private def instantiate(ty: PolyType)(using ctx: Ctx): GeneralType =
-    ty.body.subst(using (ty.tvs.map {
-      case InfVar(_, uid, _, _) =>
+  private def instantiate(ty: PolyType)(using ctx: Ctx, cctx: CCtx): GeneralType =
+    val map = (ty.tvs.map {
+      case InfVar(_, uid, state, _) =>
         val nv = freshVar
         uid -> nv
-    }).toMap)
+    }).toMap
+    ty.tvs.foreach {
+      case InfVar(_, uid, state, _) =>
+        state.upperBounds.foreach(ub => constrain(map(uid), ub.subst(using map)))
+        state.lowerBounds.foreach(lb => constrain(lb.subst(using map), map(uid)))
+    }
+    ty.body.subst(using map)
```

Comment:
We shouldn't constrain bounds on instantiation. We should assume the bounds are consistent and directly admit them as part of the fresh variable's state. The definition site where the bounds are  introduced in the first place is where consistency checking should happen, ie that all LBs are subtypes of all UBs. Is this currently done?

Reply:
The consistency check is now implemented in the function `genPolyType`.